@def title = "Home"
@def tags = ["syntax", "code"]



~~~
<div class="row">
  <div class="container">
    <img class="left" style="margin:0 15px 0 0; height:180px; width:180px;" src="/assets/adam_square.jpeg">
    <p>
       <font size="+1"><b>Adam Gayoso</b></font>
       <br>
       PhD Candidate,
       <br>
       Computational Biology,
       <br>
       <a href="http://ccb.berkeley.edu">UC Berkeley</a>,
       <br>
       adamgayoso [at] berkeley [dot] edu,
       <br>
      <a href="https://twitter.com/adamgayoso" class=" icon">
        <i class="fa fa-twitter fa-lg"></i>
      </a>
      <a href="https://github.com/adamgayoso" class="icon">
          <i class="fa fa-github fa-lg"></i>
      </a>
      <a href="https://linkedin.com/in/adam-gayoso" class=" icon">
          <i class="fa fa-linkedin fa-lg"></i>
      </a>
    </p>
    <div style="clear: both"></div>
  </div>
  <hr>
</div>

~~~

I am a third year PhD candidate in the Center for Computational Biology at UC Berkeley co-advised by [Aaron Streets](http://streetslab.berkeley.edu) and [Nir Yosef](https://niryosef.wordpress.com).
My research focuses on modeling single-cell multiomics data with deep generative models, with the goal of connecting an expressive data representation to common downstream tasks.
I am also a core developer of [single-cell variational inference tools](http://scvi-tools.org/) (scvi-tools), which is a set of probabilistic tools for analyzing single-cell data.

I received both my BS in Operations Research: Engineering Management Systems and MS in Computer Science at
Columbia University. At Columbia, I developed computational models for single-cell
RNA-sequencing data with the [Dana Pe'er Lab](https://www.mskcc.org/research-areas/labs/dana-pe-er) including a [method](https://github.com/JonathanShor/DoubletDetection) to detect doublets in single-cell RNA-sequencing datasets.

~~~
<hr>
~~~



