---
layout: page
description: "About me"
---


<div class="row" id="aboutme">
    <div class="col-md-8 offset-md-2">
        <div class="d-sm-flex justify-content-start align-items-center">
            <div class="p-2">
                <img class="rounded-circle" src="img/avatar.jpg" alt="Responsive image"
                    style="margin:0 5px 0 0; height:200px; width:200px;">
            </div>
            <div class="p-2">
                <h2>Adam Gayoso</h2>
                <h6>PhD Student,</h6>
                <h6>Computational Biology,</h6>
                <h6><a href="http://ccb.berkeley.edu">UC Berkeley</a></h6>
                <h6>adamgayoso [at] berkeley [dot] edu</h6>
                <a href="https://github.com/{{ site.github_username }}" class="icon">
                    <i class="fa fa-github fa-lg"></i>
                </a>
                <a href="https://linkedin.com/in/{{ site.linkedin_username }}"" class=" icon">
                    <i class="fa fa-linkedin fa-lg"></i>
                </a>
            </div>
        </div>
        <hr>
    </div>
    <div class="col-md-8 offset-md-2 text-left">
        <p>I am a second year graduate student in the Center for Computational Biology at UC Berkeley co-advised by <a
                href="http://streetslab.berkeley.edu">Aaron Streets</a> and <a href="https://niryosef.wordpress.com">Nir
                Yosef</a>. I am interested in integrating distinct measurements taken from single cells using deep
            generative models.</p>
        <p>
            I received my BS in Operations Research: Engineering Management Systems and MS in Computer Science at
            Columbia University. As a Columbia graduate student, I developed computational models for single-cell
            RNA-sequencing data with the <a href="https://www.mskcc.org/research-areas/labs/dana-pe-er">Dana Pe'er
                Lab</a>.
        </p>
        <h3 style="padding-top: 0.4em;">Publications</h3>
        <hr>
        <h4 style="padding-top: 0.4em; padding-bottom: 0.7em;">Preprints</h4>
        <p>
            Deep generative models for detecting differential expression in single cells.<br>
            Pierre Boyeau, Romain Lopez, Jeffrey Regier, <b>Adam Gayoso</b>, Michael I. Jordan, Nir
            Yosef.<br>
            <i>bioRxiv</i>, 2019.
            <a href="https://www.biorxiv.org/content/10.1101/794289v1">[Paper]</a>
        </p>
        <h4 style="padding-top: 0.4em; padding-bottom: 0.7em;">Journal Articles</h4>
        <p>
            Characterization of cell fate probabilities in single-cell data with Palantir.<br>
            Manu Setty, Vaidotas Kiseliovas, Jacob Levine, <b>Adam Gayoso</b>, Linas Mazutis, and Dana Peâ€™er.<br>
            <i>Nature Biotechnology</i>, 2019.
            <a href="https://www.nature.com/articles/s41587-019-0068-4">[Paper]</a>
        </p>
        <p>
            Stress-adaptive responses associated with high-level carbapenem resistance in kpc-producing klebsiella
            pneumoniae.<br>
            Sheila Adams-Sapper, <b>Adam Gayoso</b>, and Lee. W. Riley.<br>
            <i>Journal of Pathogens</i>, 2018.
            <a href="https://www.hindawi.com/journals/jpath/2018/3028290/abs/">[Paper]</a>
        </p>

        <h3 style="padding-top: 0.4em;">Software</h3>
        <hr>
        <h4 style="padding-top: 0.4em;">single-cell Variational Inference (scVI)</h4>
        <p>
            <a href="https://github.com/YosefLab/scVI">scVI</a> is an end-to-end analysis framework for single-cell
            transcriptomics.
        </p>
        <h4 style="padding-top: 0.4em;">DoubletDetection</h4>
        <p>
            <a href="https://github.com/JonathanShor/DoubletDetection">DoubletDetection</a> is a package to detect
            cross-type doublets in scRNA-seq datasets.
        </p>
    </div>
</div>